designs/SpanMarshallers.md

```diff
@@ -0,0 +1,247 @@
+
+```diff
+
+- [AttributeUsage(AttributeTargets.Parameter | AttributeTargets.ReturnValue | AttributeTargets.Field)]
++ [AttributeUsage(AttributeTargets.Parameter | AttributeTargets.ReturnValue | AttributeTargets.Field, AllowMultiple=true)]
+public class MarshalUsingAttribute : Attribute
+{
++    public MarshalUsingAttribute() {}
+     public MarshalUsingAttribute(Type nativeType) {}
++    public string CountParameterName { get; set; }
++    public int ConstantElementCount { get; set; }
++    public int ElementIndirectionLevel { get; set; }
++    public const string ReturnsCountValue = "return-value";
+}
+```
+
+The `MarshalUsingAttribute` will now provide a `CountParameterName` property that will point to a parameter whose value will hold the number of native collection elements, or to the return value if the value of `CountParameterName` is `ReturnsCountValue`. The `ConstantElementCount` property allows users to provide a constant collection length.
+
+> Open Question:
+> Should combining `CountParameterName` and `ConstantElementCount` in the same attribute be allowed?
+> With the `MarshalAs` marshalling, `SizeParamIndex` and `SizeConst` can be combined and the resulting size will be `paramValue(SizeParamIndex) + SizeConst`.
+
+To support supplying information about collection element counts, a parameterless constructor is added to the `MarshalUsingAttribute` type. The default constructor specifies that the code generator should use the information in the attribute but use the default marshalling rules for the type.
+
+The `ElementIndirectionLevel` property is added to support supplying marshalling info for element types in a collection. For example, if the user is passing a `List<List<Foo>>` from managed to native code, they could provide the following attributes to specify marshalling rules for the outer and inner lists and `Foo` separately:
+
+```csharp
+private static partial void Bar([MarshalUsing(CountParameterName = nameof(count)), MarshalUsing(ConstantElementCount = 10, ElementIndirectionLevel = 1), MarshalUsing(typeof(FooMarshaler), ElementIndirectionLevel = 2)] List<List<Foo>> foos, int count);
+```
+
+Multiple `MarshalUsing` attributes can only be supplied on the same parameter or return value if the `ElementIndirectionLevel` property is set to distinct values.
+
+Alternatively, the `MarshalUsingAttribute` could provide a `Type ElementNativeType { get; set; }` property instead of an `ElementIndrectionLevel` property and support specifying the native type of the element of the collection this way. However, this design would block support for marshalling collections of collections.
```

Comment:
>However, this design would block support for marshalling collections of collections.

I like the spirit of this, but I am dubious of its value. The interop boundary is typically not the place for complex data structures beyond arrays and matrices of types. I am all for pushing the boundaries here but it does put a fair bit of cost that I am unsure enables something real in practice. Are their existing cases where marshalling a `List<List<T>>` or perhaps a `List<Set<T>>` is warranted?

Reply:
A simple method of implementing a matrix is with a jagged array `int[][]`, which would fall into the collection-of-collections model that this section supports.

Another example would be trying to pass an `object[][]` as a SAFEARRAY-of-SAFEARRAYs-of-VARIANTs, which I don't know of any API that does, but it is enabled by the native SAFEARRAY api.

Reply:
How would `T<U<A>, V<B>>` work here? I am thinking specifically as it relates to `ElementIndirectionLevel`.